core/src/com/unciv/logic/GameStarter.kt

```diff
@@ -188,11 +189,22 @@ object GameStarter {
 
         val civNamesWithStartingLocations = gameInfo.tileMap.startingLocationsByNation.keys
 
-        val availableCityStatesNames = Stack<String>()
```

Comment:
All the confusion "just" to exclude the new CS when Religion is off... OK. Actually, looks like an almost valid approach, though it can look much simpler. Revert all that and insert a filter in the chain feeding `availableCityStatesNames.addAll`: "get all ruleset nations... limit to city states ... shuffle" can do what you need right in the "limit to" part.
```kotlin
        availableCityStatesNames.addAll(
            ruleset.nations
            .filter {
                it.value.isCityState() &&
                (it.value.cityStateType != CityStateType.Religious || newGameParameters.religionEnabled)
            }
            .keys
            .shuffled()
            .sortedByDescending { it in civNamesWithStartingLocations }
        )
```
More readable like this perhaps.

Some of your misunderstandings here:
- `val` versus `var` for a collection: No need to var that Stack - val makes the container object immutable, not its contents - that depends on the collection type. IOW, you don't want to change the toybox, you only want to change the toys in it - so the box itself can be immutable. This distinction can be hard to grasp - it's a mindset.
- Nation vs CivilizationInfo: Nation is a template, CivilizationInfo a player, a participant. This code is about building the latter using the former. The property cityStateType - yes CivilizationInfo has it but it gets it from the Nation, so while in the process of creating the requested players, such filters should go by the original.
- for..remove: Look up `ConcurrentModificationException`. A loop uses an `Iterator` - an object responsible for "where am I, what's next and when am I done". The job of an iterator becomes harder when you modify the unterlying collection while you iterate - modifications might remove what the iterator expects to be next, or similar issues. So most of them explicitly guard against modification of "their" collection. Exceptions are very specific and need a very constrained usage, so yes I don't know how, never learned those. To keep it simple, never modify a collection you or someone else may be iteration at the same time. That's why you might discover loops iterating over List.toList() - they iterate over an independent clone.

Reply:
Thanks for the help. I actually changed the val to a var in an attempt to determine if a city state is religious. I understand the difference between them.

I probably should learn more about concurrentmodificationexecptions

Thanks for the implementation

Reply:
fixed